### PR TITLE
sipsess:  added UPDATE request handler

### PR DESCRIPTION
**Pull request by juha-h**

### CODE REVIEW

**Comment by juha-h** [comment]
Should desc be mem_deref'ed before return from update_handler?

**Comment by alfredh** [comment]
one ccheck warning must be fixed:

./src/sipsess/listen.c:248: line is too wide (80 - max 79)

File: src/sipsess/listen.c [hunk 1]
```diff
@@ -245,6 +245,56 @@ static void invite_handler(struct sipsess_sock *sock,
...
+			 "Content-Length: %zu\r\n"
+			 "\r\n"
+			 "%b",
+			 sip_contact_print, &contact,
+			 desc ? "Content-Type: " : "",
+			 desc ? sess->ctype : "",
+			 desc ? "\r\n" : "",
+			 desc ? mbuf_get_left(desc) : (size_t)0,
+			 desc ? mbuf_buf(desc) : NULL,
+			 desc ? mbuf_get_left(desc) : (size_t)0);
```
Comment (alfredh) [inline src/sipsess/listen.c]: the `desc` might be allocated by offer-handler.

I guess something similar to this should be done at the end:

```
        /* pending modifications considered outdated;
           sdp may have changed in above exchange */
        sess->desc = mem_deref(sess->desc);
        sess->modify_pending = false;
        tmr_cancel(&sess->tmr);
        mem_deref(desc);
```

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> the `desc` might be allocated by offer-handler.
> 
> I guess something similar to this should be done at the end:
> 
> ```
>         /* pending modifications considered outdated;
>            sdp may have changed in above exchange */
>         sess->desc = mem_deref(sess->desc);
>         sess->modify_pending = false;
>         tmr_cancel(&sess->tmr);
>         mem_deref(desc);

Yes, I asked about it.  I'm not sure about the timer though, since
UPDATE may arrive before final response to INVITE.  Should in this case
the timer be retained?

-- Juha

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> I guess something similar to this should be done at the end:
> 
> ```
>         /* pending modifications considered outdated;
>            sdp may have changed in above exchange */
>         sess->desc = mem_deref(sess->desc);
>         sess->modify_pending = false;
>         tmr_cancel(&sess->tmr);
>         mem_deref(desc);
> ```

I freed desc, but left pending and timer as is, since there may still
exist a pending (re-)INVITE.

-- Juha

**Comment by alfredh** [comment]
do you know any other SIP clients that support UPDATE?
(except Chrome JS)

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:

> do you know any other SIP clients that support UPDATE?
> (except Chrome JS)

I first met UPDATE while testing JsSIP against baresip.  PJSIP libarary
supports UPDATE, but I don't currently have any PJSIP clients around.
Also many commercial products from Cisco, MicroSoft, etc. support
UPDATE.

-- Juha

**Comment by juha-h** [comment]
I installed pjsua and made it send UPDATE to baresip.  Below is what happened.

T 2017/06/10 18:06:17.221747 192.26.133.10:5060 -> 192.26.133.10:32816 [AP]
UPDATE sip:jh-0x5623fd7a2b00@192.168.43.98:38101;transport=tcp SIP/2.0.
Via: SIP/2.0/TCP 192.26.133.10;branch=z9hG4bK99b1.dc215024cc938edd7955d5dc12690bb2.0;i=7.
Via: SIP/2.0/TCP 192.26.133.10:53507;received=192.26.133.10;rport=53507;branch=z9hG4bKPjch7zvThg7gACFWTpP3ahOxKxhuM0mR2h;alias.
Max-Forwards: 69.
From: <sip:test@test.tutpro.com>;tag=ptrWt.X0Eh6tUPS0IHHAmhlBncpWmO5V.
To: <sip:jh@test.tutpro.com>;tag=26eb7b459a1ad566.
Contact: <sip:test@192.26.133.10:53507;transport=TCP;ob>.
Call-ID: b96c83b8e712c098.
CSeq: 6017 UPDATE.
Supported: replaces, 100rel, timer, norefersub.
Session-Expires: 1800.
Min-SE: 90.
Content-Type: application/sdp.
Content-Length:   411.
.
v=0.
o=- 3706095923 3706095928 IN IP4 192.26.133.10.
s=pjmedia.
b=AS:117.
t=0 0.
a=X-nat:0.
m=audio 30052 RTP/AVP 97 96.
c=IN IP4 192.26.133.10.
b=TIAS:96000.
a=rtpmap:97 speex/8000.
a=rtpmap:96 telephone-event/8000.
a=fmtp:96 0-16.
a=sendrecv.
a=rtcp:30053.
a=candidate:CaQPEZCXEpW0fLCJ 1 UDP 2130706431 192.26.133.10 30052 typ host.
a=candidate:CaQPEZCXEpW0fLCJ 2 UDP 2130706430 192.26.133.10 30053 typ host.

#
T 2017/06/10 18:06:17.222244 192.26.133.10:32816 -> 192.26.133.10:5060 [AP]
SIP/2.0 200 OK.
Via: SIP/2.0/TCP 192.26.133.10;branch=z9hG4bK99b1.dc215024cc938edd7955d5dc12690bb2.0;i=7.
Via: SIP/2.0/TCP 192.26.133.10:53507;received=192.26.133.10;rport=53507;branch=z9hG4bKPjch7zvThg7gACFWTpP3ahOxKxhuM0mR2h;alias.
From: <sip:test@test.tutpro.com>;tag=ptrWt.X0Eh6tUPS0IHHAmhlBncpWmO5V.
To: <sip:jh@test.tutpro.com>;tag=26eb7b459a1ad566.
Call-ID: b96c83b8e712c098.
CSeq: 6017 UPDATE.
Server: baresip v0.5.3 (x86_64/linux).
Contact: <sip:jh@test.tutpro.com;gr=urn:uuid:55860551-f714-931b-2ad9-9ca935948d8b>.
Content-Type: application/sdp.
Content-Length: 353.
.
v=0.
o=- 871727227 331734159 IN IP4 192.168.43.98.
s=-.
c=IN IP4 192.168.43.98.
t=0 0.
a=tool:baresip 0.5.3.
m=audio 10520 RTP/AVP 97 96.
b=AS:128.
a=rtpmap:97 speex/8000.
a=fmtp:97 mode="3";vbr=off;cng=on.
a=rtpmap:96 telephone-event/8000.
a=fmtp:96 0-15.
a=sendrecv.
a=label:1.
a=rtcp-rsize.
a=ssrc:646426780 cname:sip:jh@test.tutpro.com.
a=ptime:20.

**Comment by alfredh** [comment]
I have also tested it with PJSUA and it works fine to send UPDATE from pjsua to baresip.

the patch looks fine to me.

@richaas should we merge this or wait for additional review ?


/Alfred

**Comment by richaas** [comment]
The patch can not be accepted in its current form. Before we can accept an SDP offer, we need to check if there is an SDP offer/answer transaction in progress. I can think of at least three scenarios:
- we haven't answered the call yet
- we have sent a re-INVITE with an offer to which we haven't received a reply yet
- we received an empty re-INVITE, sent an offer in the the 200 response and waits for the answer in the ACK.

The patch also fails to clear pending modifications by reseting the timer and clear modify_pending flag,
sess->desc is cleared though. But none of this should be done unless the UPDATE contains SDP.

Additionally I need to study the UPDATE RFC to see if there are other scenarios we need to consider.
Haven't had the time to do that yet.

**Comment by juha-h** [comment]
Richard Aas writes:

> The patch can not be accepted in its current form. Before we can
> accept an SDP offer, we need to check if there is an SDP offer/answer
> transaction in progress. I can think of at least three scenarios:
> - we haven't answered the call yet

That is OK as long as 18x reply has been sent.

> - we have sent a re-INVITE with an offer to which we haven't received
>   a reply yet

In that case 491 reply should be sent.

> - we received an empty re-INVITE, sent an offer in the the 200
>   response and waits for the answer in the ACK.

That seems to be the same as previous.

> The patch also fails to clear pending modifications by reseting the
> timer and clear modify_pending flag,

Receiving UPDATE while waiting for final reply to INVITE does not affect
the INVITE transaction nor its timers.

> sess->desc is cleared though. But none of this should be done unless
> the UPDATE contains SDP.
> 
> Additionally I need to study the UPDATE RFC to see if there are other
> scenarios we need to consider. 
> Haven't had the time to do that yet.

OK, Juha

**Comment by richaas** [comment]
Closing due to inactivity.

**Pull request closed without merging**